File: projects/group/streamlit_app/guardian/agents/intervention_agent.py
"""Intervention Agent — picks a level (none / banner / fullScreen / delay)
and materialises an :class:`InterventionAction` with headline + body.

1:1 port of ``app/lib/agents/intervention_agent.dart``. The UI reads
``state.ambient`` for the sticky banner and ``state.pending`` for the
modal takeover.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING

from guardian.scenarios.events import (
    CallEvent,
    ChatEvent,
    ScamEvent,
    SmsEvent,
    TransactionEvent,
)

if TYPE_CHECKING:  # pragma: no cover
    from guardian.agents.context_agent import ContextSnapshot
    from guardian.agents.risk_agent import RiskAssessment


class InterventionLevel(str, Enum):
    NONE = "none"
    BANNER = "banner"
    FULL_SCREEN = "fullScreen"
    DELAY = "delay"


@dataclass(frozen=True)
class InterventionAction:
    id: str
    level: InterventionLevel
    risk: float
    headline: str
    body: str
    event_id: str
    created_at: datetime
    cooldown_seconds: int = 0
    dismissed: bool = False
    overridden: bool = False


@dataclass
class InterventionState:
    pending: InterventionAction | None = None
    ambient: InterventionAction | None = None
    history: list[InterventionAction] = field(default_factory=list)


class InterventionAgent:
    def __init__(self) -> None:
        self._state = InterventionState()
        self._counter = 0

    @property
    def state(self) -> InterventionState:
        return self._state

    def decide(self, assessment: "RiskAssessment", context: "ContextSnapshot") -> None:
        event = context.triggering_event
        level = self._level_for(assessment.final_risk, event)
        if level is InterventionLevel.NONE:
            return
        self._counter += 1
        action = InterventionAction(
            id=f"i{self._counter}",
            level=level,
            risk=assessment.final_risk,
            headline=self._headline_for(level, event),
            body=self._body_for(assessment, context),
            event_id=event.id,
            created_at=datetime.now(),
            cooldown_seconds={
                InterventionLevel.FULL_SCREEN: 60,
                InterventionLevel.DELAY: 60 * 60 * 24,
            }.get(level, 0),
        )
        history = [*self._state.history, action]
        if level is InterventionLevel.BANNER:
            self._state = InterventionState(
                pending=self._state.pending,
                ambient=action,
                history=history,
            )
        else:
            self._state = InterventionState(
                pending=action,
                ambient=self._state.ambient,
                history=history,
            )

    def dismiss_ambient(self) -> None:
        cur = self._state.ambient
        if cur is None:
            return
        new_history = [
            replace(h, dismissed=True) if h.id == cur.id else h
            for h in self._state.history
        ]
        self._state = InterventionState(
            pending=self._state.pending,
            ambient=None,
            history=new_history,
        )

    def override_pending(self) -> None:
        cur = self._state.pending
        if cur is None:
            return
        new_history = [
            replace(h, overridden=True) if h.id == cur.id else h
            for h in self._state.history
        ]
        self._state = InterventionState(
            pending=None,
            ambient=self._state.ambient,
            history=new_history,
        )

    def resolve_pending(self) -> None:
        self._state = InterventionState(
            pending=None,
            ambient=self._state.ambient,
            history=self._state.history,
        )

    def reset(self) -> None:
        self._state = InterventionState()
        self._counter = 0

    # -- helpers ---------------------------------------------------------

    @staticmethod
    def _level_for(risk: float, event: ScamEvent) -> InterventionLevel:
        is_txn = isinstance(event, TransactionEvent)
        if is_txn and risk >= 0.85:
            return InterventionLevel.DELAY
        if is_txn and risk >= 0.6:
            return InterventionLevel.FULL_SCREEN
        if risk >= 0.75:
            return InterventionLevel.FULL_SCREEN
        if risk >= 0.3:
            return InterventionLevel.BANNER
        return InterventionLevel.NONE

    @staticmethod
    def _headline_for(level: InterventionLevel, event: ScamEvent) -> str:
        if isinstance(event, CallEvent):
            subject = "this call"
        elif isinstance(event, SmsEvent):
            subject = "this message"
        elif isinstance(event, ChatEvent):
            subject = "this chat"
        elif isinstance(event, TransactionEvent):
            subject = "this transfer"
        else:
            subject = "this event"
        if level is InterventionLevel.BANNER:
            return f"Something looks off about {subject}"
        if level is InterventionLevel.FULL_SCREEN:
            return f"Pause — {subject} looks like a scam"
        if level is InterventionLevel.DELAY:
            return f"24-hour hold suggested on {subject}"
        return ""

    @staticmethod
    def _body_for(a: "RiskAssessment", c: "ContextSnapshot") -> str:
        bullets: list[str] = [f"• {reason}" for reason in a.reasons[:3]]
        if c.has_recent_call and c.seconds_since_last_call < 600:
            minutes = max(1, (c.seconds_since_last_call + 59) // 60)
            bullets.append(
                f"• You got a phone call {minutes} minute(s) ago — scammers "
                "often follow up with pressure."
            )
        if c.has_recent_sms and c.seconds_since_last_sms < 600:
            bullets.append("• A suspicious message arrived recently.")
        return "\n".join(bullets)

File: projects/group/legacy-flutter/app/lib/agents/intervention_agent.dart
import 'dart:developer' as dev;

import 'package:meta/meta.dart';
import 'package:riverpod/riverpod.dart';

import '../scenarios/events.dart';
import 'context_agent.dart';
import 'risk_agent.dart';

enum InterventionLevel { none, banner, fullScreen, delay }

@immutable
class InterventionAction {
  const InterventionAction({
    required this.id,
    required this.level,
    required this.risk,
    required this.headline,
    required this.body,
    required this.eventId,
    required this.createdAt,
    this.cooldownSeconds = 0,
    this.dismissed = false,
    this.overridden = false,
  });

  final String id;
  final InterventionLevel level;
  final double risk;
  final String headline;
  final String body;
  final String eventId;
  final DateTime createdAt;
  final int cooldownSeconds;
  final bool dismissed;
  final bool overridden;

  InterventionAction copyWith({bool? dismissed, bool? overridden}) {
    return InterventionAction(
      id: id,
      level: level,
      risk: risk,
      headline: headline,
      body: body,
      eventId: eventId,
      createdAt: createdAt,
      cooldownSeconds: cooldownSeconds,
      dismissed: dismissed ?? this.dismissed,
      overridden: overridden ?? this.overridden,
    );
  }
}

@immutable
class InterventionState {
  const InterventionState({
    this.pending,
    this.ambient,
    this.history = const [],
  });

  final InterventionAction? pending;
  final InterventionAction? ambient;
  final List<InterventionAction> history;

  InterventionState copyWith({
    InterventionAction? pending,
    bool clearPending = false,
    InterventionAction? ambient,
    bool clearAmbient = false,
    List<InterventionAction>? history,
  }) {
    return InterventionState(
      pending: clearPending ? null : (pending ?? this.pending),
      ambient: clearAmbient ? null : (ambient ?? this.ambient),
      history: history ?? this.history,
    );
  }
}

class InterventionAgent extends Notifier<InterventionState> {
  int _counter = 0;

  @override
  InterventionState build() => const InterventionState();

  Future<void> decide(
    RiskAssessment assessment,
    ContextSnapshot context,
  ) async {
    final level = _levelFor(assessment.finalRisk, context.triggeringEvent);
    if (level == InterventionLevel.none) {
      dev.log(
        '[intervention] risk=${assessment.finalRisk.toStringAsFixed(2)} → silent',
        name: 'intervention',
      );
      return;
    }
    final action = InterventionAction(
      id: 'i${++_counter}',
      level: level,
      risk: assessment.finalRisk,
      headline: _headlineFor(level, context.triggeringEvent),
      body: _bodyFor(assessment, context),
      eventId: context.triggeringEvent.id,
      createdAt: DateTime.now(),
      cooldownSeconds: switch (level) {
        InterventionLevel.fullScreen => 60,
        InterventionLevel.delay => 60 * 60 * 24,
        _ => 0,
      },
    );
    dev.log(
      '[intervention] risk=${assessment.finalRisk.toStringAsFixed(2)} → ${level.name}',
      name: 'intervention',
    );
    if (level == InterventionLevel.banner) {
      state = state.copyWith(
        ambient: action,
        history: [...state.history, action],
      );
    } else {
      state = state.copyWith(
        pending: action,
        history: [...state.history, action],
      );
    }
  }

  InterventionLevel _levelFor(double risk, ScamEvent event) {
    final isTxn = event is TransactionEvent;
    if (isTxn && risk >= 0.85) return InterventionLevel.delay;
    if (isTxn && risk >= 0.6) return InterventionLevel.fullScreen;
    if (risk >= 0.75) return InterventionLevel.fullScreen;
    if (risk >= 0.3) return InterventionLevel.banner;
    return InterventionLevel.none;
  }

  String _headlineFor(InterventionLevel level, ScamEvent event) {
    final subject = switch (event) {
      CallEvent() => 'this call',
      SmsEvent() => 'this message',
      ChatEvent() => 'this chat',
      TransactionEvent() => 'this transfer',
    };
    return switch (level) {
      InterventionLevel.banner => 'Something looks off about $subject',
      InterventionLevel.fullScreen => 'Pause — $subject looks like a scam',
      InterventionLevel.delay => '24-hour hold suggested on $subject',
      InterventionLevel.none => '',
    };
  }

  String _bodyFor(RiskAssessment a, ContextSnapshot c) {
    final bullets = <String>[];
    for (final reason in a.reasons.take(3)) {
      bullets.add('• $reason');
    }
    if (c.hasRecentCall && c.secondsSinceLastCall < 600) {
      bullets.add(
        '• You got a phone call ${(c.secondsSinceLastCall / 60).ceil()} minute(s) ago — scammers often follow up with pressure.',
      );
    }
    if (c.hasRecentSms && c.secondsSinceLastSms < 600) {
      bullets.add('• A suspicious message arrived recently.');
    }
    return bullets.join('\n');
  }

  void dismissAmbient() {
    final cur = state.ambient;
    if (cur == null) return;
    state = state.copyWith(
      clearAmbient: true,
      history: [
        for (final h in state.history)
          if (h.id == cur.id) h.copyWith(dismissed: true) else h,
      ],
    );
  }

  void overridePending() {
    final cur = state.pending;
    if (cur == null) return;
    state = state.copyWith(
      clearPending: true,
      history: [
        for (final h in state.history)
          if (h.id == cur.id) h.copyWith(overridden: true) else h,
      ],
    );
  }

  void resolvePending() {
    state = state.copyWith(clearPending: true);
  }

  void reset() {
    state = const InterventionState();
  }
}

final interventionAgentProvider =
    NotifierProvider<InterventionAgent, InterventionState>(
        InterventionAgent.new);
